### ISSUE TITLE
Custom C++ Torchscript with Composite PyTorch Functions

### ISSUE THREAD

**Issue by nickjbrowning** [pytorch/pytorch#84501]
### 🐛 Describe the bug

Hi all,

I've been trying to make a custom C++ code that I can call with torchscript, in which both the forwards and backwards pass are defined. To Illustrates this, I've created the following simple repository:

https://github.com/nickjbrowning/torchscript_test

to build this, just type:
```python
python3 setup.py build
```

and make sure the /build/lib folder is in your PYTHONPATH.

This repository provides forwards and backwards calls for f(x) = 2x, wrapped into a C++ autograd function.

Now, the gradients computed for both torch and C++ implementation  of f(x) = 2x are obviously correct, however, if I combine the torchscript implementation with a pytorch function, to yield the following:

```python
@torch.jit.script
def fn(x):
    
    xx = torch.ops.my_ops.double_op(x)
    xxx = torch.pow(x, 2)
    
    return torch.sum(xxx * xx)
```

then the derivatives of this are incorrect, while the pure pytorch implementation:

```python
test = torch.sum(torch.pow(x2, 2) * 2 * x2)
```
are correct. Here's the output of the composite torchscript-pytorch and pure-pytorch implementation, respectively, obtained from the ```double.py``` test script.

```python
input: tensor([0.3000, 0.7000], requires_grad=True)
tensor([2.3600, 3.9600])
tensor([0.5400, 2.9400])
```
Can anyone see what I'm doing wrong? 

### Versions

Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.20.2
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-46-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.4.120
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.2.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.2.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.6
[pip3] pytorch-lightning==1.6.4
[pip3] torch==1.10.0
[pip3] torch-ema==0.3
[pip3] torch-runstats==0.2.0
[pip3] torchaudio==0.10.0
[pip3] torchmetrics==0.9.0
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h8d4b97c_729    conda-forge
[conda] mkl-service               2.4.0            py39h7e14d7c_0    conda-forge
[conda] numpy                     1.21.6           py39h18676bf_0    conda-forge
[conda] pytorch                   1.10.0          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-lightning         1.6.4              pyhd8ed1ab_0    conda-forge
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-ema                 0.3                      pypi_0    pypi
[conda] torch-runstats            0.2.0                    pypi_0    pypi
[conda] torchaudio                0.10.0               py39_cu113    pytorch
[conda] torchmetrics              0.9.0              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.11.1               py39_cu113    pytorch

**Comment by ryuusama09** [comment]
hi , is this issue still open ? . IOO would like to work here

**Comment by nickjbrowning** [comment]
> hi , is this issue still open ? . IOO would like to work here

yes this is still an issue.

**Comment by shuokay** [comment]
After your code, I think the returned gradient should be `grad_output * 2`

**Comment by nickjbrowning** [comment]
this was indeed the issue, silly me!
